Add logic to clear cancelled connection requests

bsparrow435 opened the pull request:
This branch adds logic to clear cancelled connections from the
riak_core_connection_manager Pending state. I have also added a few
helper functions to give a better view into request
states(get_request_states) and connection
errors(get_connection_errrors).

Connections are cancelled only when a user runs `riak-repl disconnect
NAME/[IP,PORT]`. I’ve added logic to this path to  call
remove_cancelled_connections after waiting cm_cancellation_interval.
This interval must either be an integer or the atom never.

This logic only makes one attempt to remove the cancelled connection.
It might be nice to have a helper function that we can call to manually
clear the connections but currently this can be done by re-trying
`riak-repl disconnect`.

The riak_test showing this issue is `replication2_connections.erl` and
the issue can be demonstrated with this code by setting the
`cm_cancellation_interval` to `never`.

bsparrow435 committed:
Add logic to clear cancelled connection requests

This branch adds logic to clear cancelled connections from the
riak_core_connection_manager Pending state. I have also added a few
helper functions to give a better view into request
states(get_request_states) and connection
errors(get_connection_errrors).

Connections are cancelled only when a user runs `riak-repl disconnect
NAME/[IP,PORT]`. I’ve added logic to this path to  call
remove_cancelled_connections after waiting cm_cancellation_interval.
This interval must either be an integer or the atom never.

This logic only makes one attempt to remove the cancelled connection.
It might be nice to have a helper function that we can call to manually
clear the connections but currently this can be done by re-trying
`riak-repl disconnect`.

The riak_test showing this issue is `replication2_connections.erl` and
the issue can be demonstrated with this code by setting the
`cm_cancellation_interval` to `never`.

cmeiklejohn commented:
Dialyzer is returning the following problems:

```
riak_core_connection_mgr.erl:541: The pattern 'ok' can never match the type {'error',atom()}
```

bsparrow435 commented:
Oh god, the `connection_helper`, i hate this stupid function. It's supposed to start and then when finished hit the 'EXIT' handle_info in the riak_core_connection_mgr to iterate stats and do some other logic. However, i've seen it totally skip this when `riak_core_connection:sync_connect` returns an `{error, Reason}` which is one reason the stats module is useless. This helper function needs to be totally reworked.

With that said, the call at that line is `riak_core_connection:sync_connect` who's spec is `-spec sync_connect(ip_addr(), clientspec()) -> ok | {error, atom()}.`

I don't see the issue here.

cmeiklejohn commented:
Test for this behavior: https://github.com/basho/riak_test/pull/626

cmeiklejohn commented:
@bsparrow435 How are we looking here?

cmeiklejohn commented on src/riak_core_connection_mgr.erl:
```diff
@@ -391,12 +417,33 @@ disconnect_from_target(Target, State = #state{pending = Pending}) ->
                                                  Req#req{state = cancelled})}
     end.
 
+%% remove pending connection from state for supplied Ref.
```
Missing `@doc`.

bsparrow435 replied:
Looking at the rest of the module, most functions are missing `@doc` and type specifications. I'll add these to my new functions but would it be ok to go ahead and add the rest in this PR?

cmeiklejohn commented on src/riak_core_connection_mgr.erl:
```diff
@@ -391,12 +417,33 @@ disconnect_from_target(Target, State = #state{pending = Pending}) ->
+        false ->
+            State;
+        {value, Req, Pending2} ->
+            if Req#req.state == cancelled ->
```
Rewrite as case, if statements are terrible to read.

cmeiklejohn commented on src/riak_core_connection_mgr.erl:
```diff
@@ -391,12 +417,33 @@ disconnect_from_target(Target, State = #state{pending = Pending}) ->
 
+%% remove pending connection from state for supplied Ref.
+remove_cancelled_connection(Ref, State = #state{pending = Pending}) ->
+    case lists:keytake(Ref, #req.ref, Pending) of
```
Is this a typo `#req.ref`?  What record are you matching against here?

bsparrow435 replied:
No, it's searching the `Pending` list which is a list of `#req`'s and saying extract the entry who's `#req.ref` matches `Ref`.

cmeiklejohn commented on src/riak_core_connection_mgr.erl:
```diff
@@ -391,12 +417,33 @@ disconnect_from_target(Target, State = #state{pending = Pending}) ->
             State;
         Req ->
+            case get_cancellation_interval() of
+                never ->
```
Wouldn't `undefined` be better here?

bsparrow435 replied:
I wanted it to be explicit since unset app env's return `undefined`. With that said, the default argument in the app_helper:get_env should handle that case so i'm fine with whatever atom we want to use.

bsparrow435 committed:
Addressing PR comments

Changes the atom for canceling cleared connections from `never` to
`undefined`. Cleaned up logic and added `@doc` to one new function.

bsparrow435 commented:
Just pushed commit addressing comments. Working on associated `riak_test` now.

cmeiklejohn commented on src/riak_core_connection_mgr.erl:
```diff
@@ -391,12 +417,35 @@ disconnect_from_target(Target, State = #state{pending = Pending}) ->
 
+%% @doc remove pending connection from state for supplied Ref.
+remove_cancelled_connection(Ref, State = #state{pending = Pending}) ->
+    case lists:keytake(Ref, #req.ref, Pending) of
```
I still don't see how this will ever match, given that the second parameter is supposed to be an integer.  Does this grab the default value from the record?

bsparrow435 replied:
I did it this way because a few lines above we did the same thing in lists:keystore. My understanding is that `#req.ref` will return an integer representing it's place in the record. I'll try testing that now to confirm. 

Looking at the rest of the module and other modules in `riak_repl` this is fairly common in all `lists:key*` operations.

bsparrow435 replied:
I just confirmed this behavior with the following immediately below line 431 of this file:

```
lager:notice("BRIAN RETURN: ~p",[#req.ref]),
```

Result:

```
2014-06-04 18:50:57.294 [notice] <0.2467.0>@riak_core_connection_mgr:disconnect_from_target:431 BRIAN RETURN: 2
```

cmeiklejohn replied:
Ah, very interesting.  Thanks for taking the time to explain why it works.

reiddraper replied:
No big deal now, but that is definitely something I've never seen before. It's clever, but I think too much so.

bsparrow435 replied:
The lists and orddicts of records within the State of this module are a nightmare. It should all be simplified and requests and endpoints broken out into separate state in different modules so we can use normal methods of interacting with them. This entire module is one big list/orddict searching/editing library for the gigantic state.

cmeiklejohn replied:
I agree with @reiddraper.  It's a bit obtuse.  Can we just be explicit about the placement in the record?

bsparrow435 replied:
I went to the docs, from the bottom of the Records page: http://www.erlang.org/doc/reference_manual/records.html

```
In addition, #Record.Name returns the index in the tuple representation of Name of the record Record. Name must be an atom.
```

So this is an explicit, by design behavior.

bsparrow435 replied:
Seems the reason this works is because people were tired of defining indexes for elements in a record.

cmeiklejohn replied:
Yes, I know it's by design.  I was arguing it was obtuse, given two Erlang programmers who have been doing it for 2+ years didn't know what it was doing, so we were making an argument to just put the integer value in (or macro).

reiddraper replied:
Yep, and I know this is code you inherited, @bsparrow435, so not asking for it to be changed now. Might as well be consistent within the module until we fix the whole mess.

cmeiklejohn replied:
Just leave it as is and we'll address all of it later.

cmeiklejohn commented:
I'm still concerned given this code introduced a new Dialyzer error which is not on master which should be fixed prior to merge.  Otherwise, once the Dialyzer warning is fixed, and the test refactored, we're good to go here.

bsparrow435 commented:
I just found an issue where, if connected/disconnected in the correct order within 5 minutes, the endpoint(ex. "127.0.0.1":"10046") will be blacklisted forever. I need to add some logic to reset the retry_backoff. Will work on it tonight.

bsparrow435 commented:
Also will be adding another riak_test to show issue

bsparrow435 commented:
False alarm, i wasn't waiting long enough for the endpoint backoff to finish. Still, going to add one more step to the riak_test to prove reconnection is possible as well as adding logic to reset the backoff timer when the cancelled connection is removed.

bsparrow435 committed:
Reset endpoint backoff during connect

bsparrow435 commented:
Dialyzer issue is the last open item:

```
riak_core_connection_mgr.erl:546: The pattern 'ok' can never match the type {'error',atom()}
```

Myself and @cmeiklejohn are looking into it now.

cmeiklejohn commented:
@reiddraper found it; warning was is in the ignore file, but line number changed.  Just push a commit to update the line number.

bsparrow435 commented:
Nice find, thanks.

cmeiklejohn committed:
Update dialyzer.ignore-warnings based on new lines.

cmeiklejohn commented:
Pushed https://github.com/basho/riak_repl/commit/d61ff7ec039ac1729a1bcb985380dcc87f522a0f with updated line numbers.

cmeiklejohn commented:
:+1: d61ff7e

cmeiklejohn commented:
@borshop merge
